src/h5web/visualizations/heatmap/config.ts

```diff
@@ -0,0 +1,54 @@
+import create from 'zustand';
+import { extent } from 'd3-array';
+import { ColorMap, Domain } from './models';
+
+const EXTEND_PERCENTAGE = 0.2;
+
+export interface HeatmapConfig {
+  extendedDomain: Domain | undefined;
+  initDomains: (values: number[]) => void;
+
+  customDomain: Domain | undefined;
+  setCustomDomain: (customDomain: Domain) => void;
+
+  colorMap: ColorMap;
+  setColorMap: (colorMap: ColorMap) => void;
+
+  hasLogScale: boolean;
+  toggleLogScale: () => void;
+
+  keepAspectRatio: boolean;
+  toggleAspectRatio: () => void;
+}
+
+export const [useHeatmapConfig] = create<HeatmapConfig>(set => ({
+  extendedDomain: undefined,
+  initDomains: (values: number[]) => {
+    const [min, max] = extent(values);
+
+    if (min === undefined || max === undefined) {
+      set({ extendedDomain: undefined, customDomain: undefined });
+      return;
+    }
+
+    const extension = (max - min) * EXTEND_PERCENTAGE;
+
+    set({
+      extendedDomain: [min - extension, max + extension],
+      customDomain: [min, max], // init custom domain to data domain
+    });
+  },
```

Comment:
I think we need to take the "real" data domain `[min, max]` accessible somewhere, either stored as a variable or through a custom hook wired to `extent`. The reason is that I foresee an Auto-scale button for the limits that sets `customData` back to `dataDomain`.

PS: you didn't like my solution to keep the `dataDomain` as variable and supply a function to generate an extended version ?

Reply:
Hmm true, the `customDomain` could remain `undefined` until changed, with a fallback to the real domain. 💯 

I made the choice here to not store the real domain, which is why it made sense to store the extended domain in the config. With the change above, the extended domain computation can be moved to a custom hook. 💯